File: QD Sorting Algorithms/QD_Analysis_Improved.py
import pandas as pd
import os
import numpy as np
import glob
import re
from scipy.signal import find_peaks

# Define constants
Cs_wvl_min = 891.6
Cs_wvl_max = 898.2

# Define data folder
folder_name = r"C:\Users\Quantum Dot\Desktop\Bera Yavuz - ANC300 Movement and Images\QD_Data\11_03_2025_Test\QD_Text_File_Data_Corrected"
os.chdir(folder_name)

# List to store matching QDs with coordinates
matching_qds = []

# Process each .txt file
for filename in glob.glob('*.txt'):
    print(f"Processing: {filename}")

    # Extract [num num] from the filename
    match = re.search(r'\[(\d+)\s+(\d+)\]', filename)
    if match:
        qd_coords = f"[{match.group(1)} {match.group(2)}]"
    else:
        qd_coords = "UNKNOWN"

    # Read the full data file
    data_frame = pd.read_csv(filename, sep="\t")

    wvlngth_array = np.asarray(data_frame['Wavelen. (nm) '])
    counts_array = np.asarray(data_frame['Count Average'])

    # Find peaks
    spectrum_avg = np.mean(counts_array)
    spectrum_cutoff = 300  # Adjust as needed
    peaks, _ = find_peaks(counts_array, height=spectrum_avg + spectrum_cutoff, distance=75)

    # Check if any peaks fall in Cs transition range
    for peak in peaks:
        if Cs_wvl_min <= wvlngth_array[peak] <= Cs_wvl_max:
            print(f"QD Found in {filename} at {wvlngth_array[peak]:.3f} nm, Coordinates: {qd_coords}")
            matching_qds.append(qd_coords)
            break  # No need to check further in this file

# Output results to a text file
output_filename = "QD_Coordinates.txt"
with open(output_filename, "w") as output_file:
    for qd in matching_qds:
        output_file.write(qd + "\n")

print(f"\nQDs matching Cs D1 transition range saved to {output_filename}.")
print(f"Total matches: {len(matching_qds)}")

File: QD Sorting Algorithms/QD_Analysis_Improved.m
% QD Analysis Script 

clc; clear; close all;
warning('off');


% Define constraints
Cs_wvl_min = 894.5;
Cs_wvl_max = 894.7;
spectrum_cutoff = 1000;  % Adjust as needed

% Define data folder (containing all QD to be tested)
%folder_with_QD_tested = "C:\Users\Quantum Dot\Desktop\Bera Yavuz - ANC300 Movement and Images\QD_Data\QD_To_Analyze_A"; (Lab PC)
folder_with_QD_tested = "C:\Users\yavub\OneDrive\Desktop\QD Analysis\Corrected_Wavelengths_Raster"; % (Home pc) 

% Define the base directory
%pathway_with_QD_acquired = "C:\Users\Quantum Dot\Desktop\Bera Yavuz -ANC300 Movement and Images\QD_Data\"; (Lab pc)
main_folder = "C:\Users\yavub\OneDrive\Desktop\QD Analysis\Desirable QDs";
specific_folder = sprintf("QD_%.3f-%.3f_nm_Range_%d_Count_Up",Cs_wvl_min,Cs_wvl_max,spectrum_cutoff); %(Home PC)
pathway_with_QD_acquired = fullfile(main_folder,specific_folder);
specific_folder_text_file = fullfile(pathway_with_QD_acquired,"text_files"); 

% Check if folder exists; if not, create it
if ~exist(specific_folder_text_file, 'dir')
    mkdir(specific_folder_text_file);  
end

% Get today's date in YYYY-MM-DD format
today_date = datestr(now, 'mm_dd_yyyy');

% Create the filename with the date
textfile_name = sprintf("QD_Coordinates_%s.txt", today_date);

% Full file path
full_file_path = fullfile(pathway_with_QD_acquired, textfile_name);

files = dir(fullfile(folder_with_QD_tested, "*.txt"));

% Initialize array for matching QDs
matching_qds = [];

% Process each .txt file
for i = 1:length(files)
    filename = fullfile(folder_with_QD_tested, files(i).name);
    %fprintf("Processing: %s\n", files(i).name);
    
    % Extract [num num] from the filename
    qd_match = regexp(files(i).name, '\[(\d+)\s+(\d+)\]', 'tokens');
    if ~isempty(qd_match)
        qd_coords = sprintf("[%s %s]", qd_match{1}{1}, qd_match{1}{2});
    else
        qd_coords = "UNKNOWN";
    end
    
    % Read data
    data_table = readtable(filename);
    data_array = table2array(data_table); 

    % Convert to arrays
    wvlngth_array = data_array(:,1);
    counts_array = data_array(:,2);

    % Find peaks
    spectrum_avg = mean(counts_array);
    %spectrum_avg = 0;
    [pks, locs] = findpeaks(counts_array, 'MinPeakHeight', spectrum_avg + spectrum_cutoff, 'MinPeakDistance', 75);

    % Check if any peaks fall in Cs transition range
    if any(wvlngth_array(locs) >= Cs_wvl_min & wvlngth_array(locs) <= Cs_wvl_max)
        fprintf("QD Found in %s at %.3f nm, Coordinates: %s\n", files(i).name, wvlngth_array(locs(1)), qd_coords);
        matching_qds = [matching_qds; qd_coords];  % Append to list
        % Copy the file
        copyfile(filename, specific_folder_text_file);
    end

    % Progression message
    fprintf("Completed Reading: %d/%d\n",i,length(files))

    
end

% Output results to a text file
fid = fopen(full_file_path, 'w');
for j = 1:length(matching_qds)
    fprintf(fid, "%s\n", matching_qds(j));
end
fclose(fid);

fprintf("\nQDs matching Cs D1 transition range saved to %s.\n", full_file_path);
disp("*********************************************************************************")
fprintf("Total matches: %d\n", length(matching_qds));
disp("*********************************************************************************")

